### PR TITLE
[AutoDiff] Support differentiation of non-active `try_apply`.

### PR DESCRIPTION
**Pull request by efremale**
Add differentiation support for non-active `try_apply` SIL instructions.

Notable pullback generation changes:
* Original basic blocks are now visited in a different order: starting from the original exit block, all its predecessors
   are visited in a breadth-first search order.
   This ensures that all successors of any block are visited before the block itself.

Resolves https://bugs.swift.org/browse/TF-433.

Examples of supported code:

```
func throwing() throws -> Void {}

@differentiable
func throwingIdentity(_ x: Float) -> Float {
  try! throwing() // non-active `try_apply`
  return x
}
```

```
func rethrowing(_ body: () throws -> Void) rethrows -> Void {}

@differentiable
func rethrowingIdentity(_ x: Float) -> Float {
  rethrowing({}) // non-active `try_apply`
  return x
}
```

### CODE REVIEW

**Comment by efremale** [comment]
cc @dan-zheng

File: lib/SILOptimizer/Differentiation/PullbackCloner.cpp [hunk 1]
```diff
@@ -1722,21 +1722,28 @@ bool PullbackCloner::Implementation::run() {
...
-    auto *origBB = origNode->getBlock();
-    postDomOrder.pushChildren(origNode);
-    // If node is the `nullptr` marker basic block, do not push it.
-    if (!origBB)
-      continue;
-    postOrderPostDomOrder.push_back(origBB);
+  // dominance order.
+  SmallVector<SILBasicBlock *, 8> dominanceOrder = {};
+  std::deque<SILBasicBlock *> queue = {};
+  SmallDenseSet<SILBasicBlock *, 8> visitedBlocks = {};
```
Comment (dan-zheng) [inline lib/SILOptimizer/Differentiation/PullbackCloner.cpp:1728]: Minor naming and style suggestions:
```suggestion
  // The original blocks in traversal order for pullback generation.
  SmallVector<SILBasicBlock *, 8> originalBlocks;
  // The set of visited original blocks.
  SmallDenseSet<SILBasicBlock *, 8> visitedOriginalBlocks;
  
  // Perform BFS from the original exit block.
  {
    std::deque<SILBasicBlock *> worklist;
    ...
  }
```

File: lib/SILOptimizer/Differentiation/PullbackCloner.cpp [hunk 2]
```diff
@@ -1722,21 +1722,28 @@ bool PullbackCloner::Implementation::run() {
   }
 
   // Create pullback blocks and arguments, visiting original blocks in
-  // post-order post-dominance order.
-  SmallVector<SILBasicBlock *, 8> postOrderPostDomOrder;
-  // Start from the root node, which may have a marker `nullptr` block if
-  // there are multiple roots.
-  PostOrderPostDominanceOrder postDomOrder(postDomInfo->getRootNode(),
-                                           postOrderInfo, original.size());
```
Comment (dan-zheng) [inline lib/SILOptimizer/Differentiation/PullbackCloner.cpp:1730]: Can you please delete `PostOrderPostDominanceOrder` (a differentiation-specific utility), now that it's no longer used?

File: lib/SILOptimizer/Differentiation/PullbackCloner.cpp [hunk 3]
```diff
@@ -1722,21 +1722,28 @@ bool PullbackCloner::Implementation::run() {
...
-  while (auto *origNode = postDomOrder.getNext()) {
-    auto *origBB = origNode->getBlock();
-    postDomOrder.pushChildren(origNode);
-    // If node is the `nullptr` marker basic block, do not push it.
-    if (!origBB)
-      continue;
-    postOrderPostDomOrder.push_back(origBB);
+  // dominance order.
```
Comment (dan-zheng) [inline lib/SILOptimizer/Differentiation/PullbackCloner.cpp:1725]: I don't think ["dominance order"](https://en.wikipedia.org/wiki/Dominator_(graph_theory)) is an accurate description for the new traversal order. Dominance order traversal starts at the function entry basic block and continues to successors based on dominance information.

"BFS from the original exit block" seems like an exactly accurate description. How about:

```suggestion
  // Create pullback blocks and arguments, visiting original blocks using BFS
  // starting from the original exit block. Unvisited original basic blocks
  // (e.g unreachable blocks) are not relevant for pullback generation and thus
  // ignored.
```

**Review by dan-zheng — APPROVED** [review]
The basic block traversal order simplification makes control flow differentiation more robust!

**Review by rxwei — APPROVED** [review]
Very nice!

**Comment by dan-zheng** [comment]
@swift-ci Please test

**Comment by dan-zheng** [comment]
@swift-ci Please test Linux

**Pull request merged**